uniffi_bindgen/src/bindings/swift/gen_swift/enum_.rs

```diff
@@ -0,0 +1,94 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::fmt;
+
+use crate::bindings::backend::{CodeDeclaration, CodeOracle, CodeType, Literal};
+use crate::interface::{ComponentInterface, Enum};
+use askama::Template;
+
+use super::filters;
+pub struct EnumCodeType {
+    id: String,
+}
+
+impl EnumCodeType {
+    pub fn new(id: String) -> Self {
+        Self { id }
+    }
+}
+
+impl CodeType for EnumCodeType {
+    fn type_label(&self, oracle: &dyn CodeOracle) -> String {
+        oracle.class_name(&self.id)
+    }
+
+    fn canonical_name(&self, oracle: &dyn CodeOracle) -> String {
+        format!("Enum{}", self.type_label(oracle))
+    }
+
+    fn literal(&self, oracle: &dyn CodeOracle, literal: &Literal) -> String {
+        if let Literal::Enum(v, _) = literal {
+            format!(".{}", oracle.enum_variant_name(v))
+        } else {
+            unreachable!();
+        }
+    }
+
+    fn lower(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display) -> String {
+        format!("{}.lower()", oracle.var_name(nm))
+    }
+
+    fn write(
+        &self,
+        oracle: &dyn CodeOracle,
+        nm: &dyn fmt::Display,
+        target: &dyn fmt::Display,
+    ) -> String {
+        format!("{}.write(into: {})", oracle.var_name(nm), target)
+    }
+
+    fn lift(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display) -> String {
+        format!("{}.lift({})", self.type_label(oracle), nm)
+    }
+
+    fn read(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display) -> String {
+        format!("{}.read(from: {})", self.type_label(oracle), nm)
+    }
+
+    fn helper_code(&self, oracle: &dyn CodeOracle) -> Option<String> {
+        Some(format!(
+            "// Helper code for {} enum is found in EnumTemplate.swift",
+            self.type_label(oracle)
+        ))
+    }
```

Comment:
I'll just add this here, for want of a better place to put it: for Swift and Rust you can add traits to structs (and classes).

Rust has rules around owning either the trait or the struct (the orphan rule), but Swift does not. There are rules around visibility of the extensions, but Swift lets you add extensions to classes, structs, and protocols *in place*. Thus:

```swift
let obj = 1
let isInt = obj is Int // true

protocol Labelled {
    var label: String { get }
}
extension Int : Labelled {
    var label: String { "Int" }
}

let isLabelled = obj is Labelled // true
let label = obj.label // "Int"
```

which means we can write fairly simple extensions for structural types:

```swift
extension Optional : Labelled where Wrapped: Labelled {
    var label: String {
        guard let inner = self?.label else {
            return "nil"
        }
        return "\(inner)?"
    }
}

let obj: Int? = 1
let isLabelled = obj is Labelled // true
let label = obj.label // "Int?"
```

This is why everything can call the `write` method without going through the `write_swift` filter, it's all fairly homogenous. There are reasons why this isn't great; but on the absenting these read/write/lift/lower probably don't need to go through a `CodeType`.

For Kotlin, on the other hand, you can add extension methods to classes and interfaces, but they're syntactic sugar for static functions. You cannot add an extension interface to an existing class.

```kotlin
interface Labelled {
    fun label(): String
}

fun Int.label() = "Int"

val obj: Int = 1
val isLabelled = obj is Labelled // false
val label = obj.label() // "Int"
```
Conceivably, we could make sure every type has the right methods (in this case `label`), but then when it came to writing structural types we can't use:

```kotlin
fun <T: Labelled> T?.label() = this?.label ?: "null"

val obj: Int? = 2
val label = obj.label() // compile error!
```

because none of the types the user is using is implementing the `Labelled` protocol, even though it implements everything it needs to.

This of course, sucks and means that we need some logic around calling of the lift/lower/read/write methods for _at least_ the Kotlin backend, and other languages that don't have extension protocols or traits.

If we now have to have that logic for Kotlin, why do we keep with it for Swift or any other language:

1. uniformity/regularity. We want to be able to have the backends to all look and work approximately the same. The code it generates may not be the same, but if each backend is its own special snowflake, then we hinder our development velocity.
2. visibility rules. These are different in each language, and I think we should try and stay out of the way of client code as much as possible. Any runtime code that we add should not be visible to the the client code. For Swift, I'm not sure if `ViaFfi` is visible to the client code or not (explicitly you have to make it public, even if all the methods are private).
3. naming collisions. We do no explicit checking on this right now: uniffi will generate invalid code in arbitrary ways if the UDL code uses methods of the same names as used by the lift/lower/read/write/coerce machinery of a type. There are ways around it, but splitting off that machinery (e.g. in our`CallbackInterface` kotlin backend we use `CallbackInterfaceInternals`) from the type itself.

This comment should probably go someplace more durable that a PR review comment. :)

Reply:
I have a PR coming along these lines.  I think the solution for Kotlin is the same solution for Rust external types, we separate the `FFIConverter` type from the type that it's converting.  So you define objects like this:

```
object FFIConverterOptionalTypeFoo: FFIConverter<Foo?, RustBuffer> {
  fun lift(RustBuffer): Foo? { ... }
  fun lower(Foo?): RustBuffer { ... }
}
```

Then you add some logic to the template code so that it can map each type to the `FFIConverter` instance and you're there.

I think something like this is needed if you're going allow users to wrap types.  They need some way to hook in to the existing lift/lower logic and if that only exists in the template code, then it's going to be hard.

Reply:
I'm definitely on board for converter objects implementing a `FFIConverter<Foo?, RustBuffer>`, for both Kotlin and Swift.